### ISSUE TITLE
I want to get F1 score in classification model

### ISSUE THREAD

**Issue by HurairaCODE** [ultralytics/ultralytics#3875]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hey, I want to know how can I get f1 score for classification model in YOLO, I have been stuck at this and will be very grateful if you can help me!

**Comment by glenn-jocher** [comment]
@HurairaCODE hello! Great question, happy to help. 

The F1 score for classification in YOLOv8 is not directly calculated within the codebase, as the primary evaluation metric in YOLOv8 is mAP (mean average precision) used for object detection tasks. However, you can certainly calculate the F1 score on your own from the precision and recall values that are provided by the model during validation.

Remember, the F1 score is the harmonic mean of precision and recall. You can access the precision and recall values from the outputs of the model validation. These values will be in the '.val' file in your runs/train directory.

Once you have these values (precision, recall), calculating F1 score would then be a straightforward case of inserting them into the F1 score formula:

F1 Score = 2 * (Precision * Recall) / (Precision + Recall)

This gives equal weight to Precision and Recall in the resulting score.

I hope this helps, and let me know if you have any further questions!

**Comment by saikrishna0203** [comment]
@glenn-jocher  please help bro i need to find precision and f1 square and also false points and positive points,i need your help i am using yolov8 model,which file i have to modify and what is code i have to write

**Comment by glenn-jocher** [comment]
@saikrishna0203 hello, thank you for using YOLOv8 and reaching out with your question.

In YOLOv8, these metrics such as precision, F1 score, false positives, and true positives are not directly computed within the codebase. The primary metrics used by the model during the validation stage are Precision and Recall, from which you could compute the F1 score manually afterwards.

The Precision and Recall computations are carried out by the model during validation and are saved to the '.val' file in your runs/train directory. Once you have these metrics, you can calculate the F1 score using the formula:

F1 Score = 2 * (Precision * Recall) / (Precision + Recall)

To find the number of False Positives and True Positives, you would need to analyse the prediction outputs of the model, comparing the predicted labels with your ground-truth labels. True positives are cases where the model correctly identified the object, and false positives are cases where the model incorrectly predicted the presence of an object.

Please note that being an open-source project, the codebase of YOLOv8 allows for modification and tweaking to fit your specific needs, so you can add these calculations directly into the model code if required.

I hope this provides some clarity. Feel free to ask if you have any further doubts.

**Comment by saikrishna0203** [comment]
Sir I need code sir

**Comment by glenn-jocher** [comment]
@saikrishna0203 hello, thank you for your question. 

The precision and recall metrics in YOLOv8 are calculated during the validation stage and are stored in a '.val' file in your runs/train directory. But based on your message, it seems like you're only seeing output images.

This could potentially indicate that you're missing the `.val` file or it's not being generated correctly during your training process. Please ensure you are executing the validation phase correctly after your training phase, as the validation step is crucial for generating these metrics.

By reviewing prediction outputs of your model and comparing these predicted labels against the true labels from your dataset, you should be able to calculate true positives (where the model correctly identifies an object) and false positives (where the model incorrectly predicts the presence of an object).

Remember that you can calculate the F1 score manually using the Precision and Recall metrics. The formula is `F1 Score = 2 * (Precision * Recall) / (Precision + Recall)`. 

Please note, that although these metrics aren't computed directly within the YOLOv8 codebase, it does provide an allowance for modification to fit your specific needs. So, you could insert these calculations directly into the model code if required.

I hope this helps. If you have any more questions, feel free to reach out.

**Comment by saikrishna0203** [comment]
Sir how to get labels from validation set pls tell me that small code and
true labels pls

**Comment by glenn-jocher** [comment]
@saikrishna0203 Hello! Accessing labels from the validation set in YOLOv8 typically involves the DataLoader that loads the validation dataset. The DataLoader is used during both the training and validation phases, and it loads batches of image-label pairs for processing by the model.

The labels are loaded along with the corresponding images into memory each time a batch is fed into the model during the validation stage. Each label typically includes information about the object category and the coordinates of the bounding box.

To access these labels, you could modify the validation phase of your model script to output or save them along with the respective image IDs. 

Keep in mind that you need to make sure that your validation set is properly annotated with the corresponding labels in the correct format as required by the YOLOv8 architecture. This is crucial for the evaluation metrics to be computed accurately.

I hope this helps. Feel free to ask if you have any further questions!

**Comment by DecipherData** [comment]
I am using YOLOv8 for image classification. I do not have annotated data as I am not using it for detection. My images are contained in the respective 11 category folders in each of train, test, and val folders. Yolov8 -nano and small pretrained models are used for classification. I want classification report and confusion matrix. Confusion matrix is given as a png file only. Only labeled and predicted images are present. There is no .val file in /runs/train. I want precision, recall scores, and F1 scores for comparisons. These are the most crucial for classification problem. Kindly help.

**Comment by glenn-jocher** [comment]
@DecipherData i understand that you're working on an image classification task using YOLOv8 and require a classification report including precision, recall, and F1 scores. In YOLOv8 for classification tasks, you might need to implement additional code to compute and display these metrics as they are more traditional within the context of the detection framework that YOLO provides.

Typically with YOLO, if you're performing pure classification (without localization), you may have to adapt the model output to suit the metrics that you mentioned. If YOLOv8 is not providing the detailed classification report out-of-the-box, you'll need to extract the predicted class probabilities from the model's output and the true class labels from your dataset. 

You should write custom code that:
1. Runs your validation dataset through the model to get predictions.
2. Compares the predictions with the true labels to identify true positives, false positives, and false negatives.
3. Calculates precision, recall, and F1 score for each class and overall.

For generating a confusion matrix programmatically, you can use libraries like Scikit-learn which provide utilities like `confusion_matrix` and `classification_report` for these tasks. Given a set of true labels and predicted labels, these functions can calculate and display the metrics you're interested in.

Please consider writing a script that does the following:
- Loads the trained model and validation dataset.
- Performs inference on the entire validation set.
- Captures the model outputs and true class labels.
- Uses those outputs with Scikit-learn's functionality to compute desired metrics.

Remember to handle the YOLO model's output appropriately, as it might be in a form suited for detection rather than flat classification and might require you to extract the class-specific prediction probabilities. 

I hope this information aids in progressing with your classification task.

**Comment by DecipherData** [comment]
> > I am using YOLOv8 for image classification. I do not have annotated data as I am not using it for detection. My images are contained in the respective 11 category folders in each of train, test, and val folders. Yolov8 -nano and small pretrained models are used for classification. I want classification report and confusion matrix. Confusion matrix is given as a png file only. Only labeled and predicted images are present. There is no .val file in /runs/train. I want precision, recall scores, and F1 scores for comparisons. These are the most crucial for classification problem. Kindly help as it is urgently required.
> 
> Hey @DecipherData did you find any way for it?

Yes, had to code for getting the classification report. You may want to extract predictions from the results. for example after loading your model, make predictions on the test images and store in results [ results = model(source=/path/to/test_images_folder , conf=0.25)  ] and using your ground truth classes, can use scikit-learn classification report package for calculating classification report. I hope it helps.

**Comment by glenn-jocher** [comment]
Hey @DecipherData! 😊

It sounds like you've made some progress! For generating a classification report and confusion matrix, you're right; leveraging scikit-learn's utilities is a great approach. Here's a quick example on how you might do it:

```python
from sklearn.metrics import classification_report, confusion_matrix
import numpy as np

# Assuming 'predictions' and 'true_labels' are your model's predictions and the actual labels respectively
predictions = np.array([result.prediction for result in results])
true_labels = np.array([your_method_to_get_true_label(result) for result in results])

print(classification_report(true_labels, predictions))
print(confusion_matrix(true_labels, predictions))
```

Make sure to replace `your_method_to_get_true_label` with your actual method for fetching the true labels from your dataset. This should give you a detailed classification report and confusion matrix for your model's performance on the test set.

Hope this helps! Keep pushing forward! 🚀

**Comment by shivam55sit** [comment]
For printing the classification report and confusion matrix for the test set images, we can manually write the code based on the predictions and actual labels of our images. I also trained a multi class image classification problem using pre-trained nano classification model of YOLOv8.

`#code for printing the classification report and confusion matrix

actual = [] #list to add actaual class labels
predicted = [] #list to add predicted class

# model_main = YOLO("/home/shivam/Desktop/non_negative_images/runs/classify/train/weights/[best.pt](http://best.pt/)")
## model_main is the best model trained on training data

dir_main = "/home/shivam/Desktop/non_negative_images/corneal_infection/test"

for path in os.listdir(dir_main):
    dir = os.path.join(dir_main,path)
    for images in os.listdir(dir):
        img_path = os.path.join(dir,images)
        # print(img_path)

        results = model_main.predict(img_path)[0]

        actual.append(results.path.split('/')[7])

        idx_max=np.argmax(results.probs.data)

        if idx_max==0:
            predicted.append('GNB')
        elif idx_max==1:
            predicted.append('GNC')
        elif idx_max==2:
            predicted.append('GPB')
        else:
            predicted.append('GPC')`

`from sklearn.metrics import classification_report, confusion_matrix

#Confusion matrix
cm = confusion_matrix(actual, predicted)

print(cm)

#Classification report

print(classification_report(actual,predicted)

 `

**Comment by shivam55sit** [comment]
For printing the classification report and confusion matrix for the test set images, we can manually write the code based on the predictions and actual labels of our images. I also trained a multi class image classification problem using pre-trained nano classification model of YOLOv8.

`# code for printing the classification report and confusion matrix

actual = [] # list to add actual class labels
predicted = [] # list to add predicted class

#  model_main = YOLO("/home/shivam/Desktop/non_negative_images/runs/classify/train/weights/best.pt")
# model_main is the best model trained on training data

dir_main = "/home/shivam/Desktop/non_negative_images/corneal_infection/test"

for path in os.listdir(dir_main):
    dir = os.path.join(dir_main,path)
    for images in os.listdir(dir):
        img_path = os.path.join(dir,images)
        # print(img_path)

        results = model_main.predict(img_path)[0]

        actual.append(results.path.split('/')[7])

        idx_max=np.argmax(results.probs.data)

        if idx_max==0:
            predicted.append('GNB')
        elif idx_max==1:
            predicted.append('GNC')
        elif idx_max==2:
            predicted.append('GPB')
        else:
            predicted.append('GPC')

print(classification_report(true_labels, predictions))
print(confusion_matrix(true_labels, predictions))
`

**Comment by shivam55sit** [comment]
For printing the classification report and confusion matrix for the test set images, we can manually write the code based on the predictions and actual labels of our images. I also trained a multi class image classification problem using pre-trained nano classification model of YOLOv8.

```
#code for printing the classification report and confusion matrix

actual = [] #list to add actaual class labels
predicted = [] #list to add predicted class

#model_main = YOLO("/home/shivam/Desktop/non_negative_images/runs/classify/train/weights/[best.pt](http://best.pt/)")
##model_main is the best model trained on training data

dir_main = "/home/shivam/Desktop/non_negative_images/corneal_infection/test"

for path in os.listdir(dir_main):
    dir = os.path.join(dir_main,path)
    for images in os.listdir(dir):
        img_path = os.path.join(dir,images)
        # print(img_path)

        results = model_main.predict(img_path)[0]

        actual.append(results.path.split('/')[7])

        idx_max=np.argmax(results.probs.data)

        if idx_max==0:
            predicted.append('GNB')
        elif idx_max==1:
            predicted.append('GNC')
        elif idx_max==2:
            predicted.append('GPB')
        else:
            predicted.append('GPC')


print(classification_report(actual, predicted))
print(confusion_matrix(actual, predicted))
```

**Comment by pderrenger** [comment]
Thank you for sharing your approach. To generate a classification report and confusion matrix for your YOLOv8 classification model, you can indeed use the predictions and actual labels from your test set. Your method of iterating through the test images, making predictions, and then comparing them to the true labels is correct. 

If you encounter any issues or bugs, please ensure you are using the latest version of the Ultralytics package. If the problem persists, feel free to provide more details, and we can look into it further. 

For any further assistance, please continue to use this platform. Thank you for your understanding.

**Comment by shivam55sit** [comment]
Thank you for the clarification.

**Comment by pderrenger** [comment]
Thank you for your detailed explanation. Your approach to generating a classification report and confusion matrix by iterating through test images and comparing predictions to true labels is indeed correct. If you encounter any issues, please ensure you are using the latest version of the Ultralytics package. If problems persist, provide more details, and we will assist you further. For any additional support, please continue to use this platform. Thank you.

**Comment by HalkScout** [comment]
The F1 score is calculated in [metrics.py](https://github.com/ultralytics/ultralytics/blob/db4c43bafbed1e1298c4436b8c7e7c6e5b9f3eec/ultralytics/utils/metrics.py#L621). 

Here was my implementation:

`utils/metrics.py`
```diff
@@ -694,6 +694,16 @@ def mr(self):
         """
         return self.r.mean() if len(self.r) else 0.0
 
+    @property
+    def mf1(self):
+        """
+        Returns the Mean Recall of all classes.
+
+        Returns:
+            (float): The mean recall of all classes.
+        """
+        return self.f1.mean() if len(self.f1) else 0.0
+
     @property
     def map50(self):
         """
@@ -726,11 +736,11 @@ def map(self):
 
     def mean_results(self):
         """Mean of results, return mp, mr, map50, map."""
-        return [self.mp, self.mr, self.map50, self.map]
+        return [self.mp, self.mr, self.map50, self.map, self.mf1]
 
     def class_result(self, i):
         """Class-aware result, return p[i], r[i], ap50[i], ap[i]."""
-        return self.p[i], self.r[i], self.ap50[i], self.ap[i]
+        return self.p[i], self.r[i], self.ap50[i], self.ap[i], self.f1[i]
 
     @property
     def maps(self):
@@ -742,7 +752,7 @@ def maps(self):
 
     def fitness(self):
         """Model fitness as a weighted combination of metrics."""
-        w = [0.0, 0.0, 0.1, 0.9]  # weights for [P, R, mAP@0.5, mAP@0.5:0.95]
+        w = [0.0, 0.0, 0.1, 0.9, 0.0]  # weights for [P, R, mAP@0.5, mAP@0.5:0.95]
         return (np.array(self.mean_results()) * w).sum()
 
     def update(self, results):
@@ -850,7 +860,7 @@ def process(self, tp, conf, pred_cls, target_cls):
     @property
     def keys(self):
         """Returns a list of keys for accessing specific metrics."""
-        return ["metrics/precision(B)", "metrics/recall(B)", "metrics/mAP50(B)", "metrics/mAP50-95(B)"]
+        return ["metrics/precision(B)", "metrics/recall(B)", "metrics/mAP50(B)", "metrics/mAP50-95(B)", "metrics/F1(B)"]
 
     def mean_results(self):
         """Calculate mean of detected objects & return precision, recall, mAP50, and mAP50-95."""
```

`models/yolo/<yolo_model>/val.py`
```diff
@@ -82,7 +82,7 @@ def init_metrics(self, model):
 
     def get_desc(self):
         """Return a formatted string summarizing class metrics of YOLO model."""
-        return ("%22s" + "%11s" * 6) % ("Class", "Images", "Instances", "Box(P", "R", "mAP50", "mAP50-95)")
+        return ("%22s" + "%11s" * 7) % ("Class", "Images", "Instances", "Box(P", "R", "mAP50", "mAP50-95", "F1)")
```

**Comment by ManuelZ** [comment]
```
metrics = model.val(split="test")

# +-------------+---------------------+
# | Predicted   | Positive | Negative |
# +-------------+---------------------+
# | Actual      |          |          |
# | Positive    |    TP    |    FP    |
# | Negative    |    FN    |    TN    |
# +-------------+---------------------+

m = metrics.confusion_matrix.matrix
tp = m[0, 0]
tn = m[1, 1]
fn = m[1, 0]
fp = m[0, 1]

f1 = 2 * tp / (2 * tp + fp + fn)
precision = tp / (tp + fp)
recall = tp / (tp + fn)

print(f"Model: '{model_path.name}'")
print(f"  - F1 score: {f1:.3f}")
print(f"  - Precision score: {precision:.3f}")
print(f"  - Recall score: {recall:.3f}")
print(f"  - Accuracy score: {metrics.top1:.3f}")
```

**Comment by glenn-jocher** [comment]
@ManuelZ thank you for sharing your code snippet! It looks great for calculating precision, recall, F1 score, and accuracy from the confusion matrix generated during validation. If you face any issues or need further clarification, feel free to ask. For more details on the validation process and metrics interpretation, you can also refer to the [Ultralytics YOLO docs on model testing](https://docs.ultralytics.com/guides/model-testing/).
